File: frontend/src/stores/xslt_templates.ts
import { defineStore } from "pinia";
import { ref } from "vue";
import { apiClient } from "@/services/api";

export interface XsltTemplateSummary {
  id: string;
  name: string;
  description: string | null;
  processor: string;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface XsltTemplateDetail extends XsltTemplateSummary {
  content: string;
  created_by: string | null;
}

export interface XsltTemplateCreate {
  name: string;
  description?: string | null;
  content: string;
  processor?: string;
  tags?: string[];
}

export interface XsltTemplatePatch {
  name?: string;
  description?: string | null;
  content?: string;
  processor?: string;
  tags?: string[];
}

export const useXsltTemplateStore = defineStore("xslt_templates", () => {
  const templates = ref<XsltTemplateSummary[]>([]);

  async function fetchTemplates(): Promise<void> {
    templates.value = await apiClient.get<XsltTemplateSummary[]>("/xslt-templates");
  }

  async function getTemplate(id: string): Promise<XsltTemplateDetail> {
    return apiClient.get<XsltTemplateDetail>(`/xslt-templates/${id}`);
  }

  async function createTemplate(payload: XsltTemplateCreate): Promise<XsltTemplateDetail> {
    const tpl = await apiClient.post<XsltTemplateDetail>("/xslt-templates", payload);
    templates.value.push(tpl);
    templates.value.sort((a, b) => a.name.localeCompare(b.name));
    return tpl;
  }

  async function patchTemplate(id: string, payload: XsltTemplatePatch): Promise<XsltTemplateDetail> {
    const tpl = await apiClient.patch<XsltTemplateDetail>(`/xslt-templates/${id}`, payload);
    const idx = templates.value.findIndex((t) => t.id === id);
    if (idx !== -1) templates.value[idx] = tpl;
    return tpl;
  }

  async function deleteTemplate(id: string): Promise<void> {
    await apiClient.delete<void>(`/xslt-templates/${id}`);
    templates.value = templates.value.filter((t) => t.id !== id);
  }

  return { templates, fetchTemplates, getTemplate, createTemplate, patchTemplate, deleteTemplate };
});

File: backend/app/routers/xslt_templates.py
"""Router for the XSLT stylesheet catalog.

All endpoints are restricted to Designer+ (Designer, EditorInChief, Admin).
The list endpoint also serves the website Document tab so it returns
XsltTemplateSummary (no full content) for bandwidth efficiency.
The detail endpoint returns the full XsltTemplateResponse including content.
"""

import uuid
from pathlib import Path
from typing import Annotated

from fastapi import APIRouter, Depends, Request
from fastapi.responses import FileResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.exceptions import AuthorizationError
from app.db.postgres import get_async_session
from app.core.constants import ROLE_LEVEL
from app.middleware.acl import get_current_user
from app.models.user import User
from app.schemas.common import DataResponse
from app.schemas.xslt_templates import (
    XsltTemplateCreate,
    XsltTemplatePatch,
    XsltTemplateResponse,
    XsltTemplateSummary,
)
from app.services.xslt_templates import (
    create_xslt_template,
    delete_xslt_template,
    get_xslt_template,
    list_xslt_templates,
    patch_xslt_template,
)

router = APIRouter(prefix="/xslt-templates", tags=["xslt-templates"])

_DEFAULT_XSLT_PATH = Path(__file__).parent.parent / "xslt" / "tei_generic.xsl"


async def _designer_plus(
    user: Annotated[User, Depends(get_current_user)],
    request: Request,
) -> User:
    """[D+] — Designer, EditorInChief, or Admin."""
    role: str = getattr(request.state, "role", "User")
    if role != "Designer" and ROLE_LEVEL.get(role, 0) < 3:
        raise AuthorizationError()
    return user


_D = Depends(_designer_plus)


@router.get("")
async def xslt_templates_list(
    current_user: Annotated[User, _D],
    db: Annotated[AsyncSession, Depends(get_async_session)],
) -> DataResponse[list[XsltTemplateSummary]]:
    """Return all XSLT templates (summary — no content body)."""
    templates = await list_xslt_templates(db)
    return DataResponse(data=[XsltTemplateSummary.model_validate(t) for t in templates])


@router.get("/default/download")
async def xslt_template_default_download(
    current_user: Annotated[User, _D],
) -> FileResponse:
    """Download the built-in TEI→HTML stylesheet (tei_generic.xsl) as a file."""
    return FileResponse(
        path=str(_DEFAULT_XSLT_PATH),
        media_type="application/xml",
        filename="tei_generic.xsl",
    )


@router.get("/{template_id}")
async def xslt_template_detail(
    template_id: uuid.UUID,
    current_user: Annotated[User, _D],
    db: Annotated[AsyncSession, Depends(get_async_session)],
) -> DataResponse[XsltTemplateResponse]:
    """Return a single XSLT template including the full stylesheet content."""
    tpl = await get_xslt_template(db, template_id)
    return DataResponse(data=XsltTemplateResponse.model_validate(tpl))


@router.post("", status_code=201)
async def xslt_template_create(
    payload: XsltTemplateCreate,
    current_user: Annotated[User, _D],
    db: Annotated[AsyncSession, Depends(get_async_session)],
) -> DataResponse[XsltTemplateResponse]:
    tpl = await create_xslt_template(db, payload, created_by=current_user.id)
    await db.commit()
    return DataResponse(data=XsltTemplateResponse.model_validate(tpl))


@router.patch("/{template_id}")
async def xslt_template_patch(
    template_id: uuid.UUID,
    payload: XsltTemplatePatch,
    current_user: Annotated[User, _D],
    db: Annotated[AsyncSession, Depends(get_async_session)],
) -> DataResponse[XsltTemplateResponse]:
    tpl = await patch_xslt_template(db, template_id, payload)
    await db.commit()
    return DataResponse(data=XsltTemplateResponse.model_validate(tpl))


@router.delete("/{template_id}", status_code=204)
async def xslt_template_delete(
    template_id: uuid.UUID,
    current_user: Annotated[User, _D],
    db: Annotated[AsyncSession, Depends(get_async_session)],
) -> None:
    await delete_xslt_template(db, template_id)
    await db.commit()
